### PR TITLE
Work 289 v2

### PR DESCRIPTION
**Pull request by hsuchie4**
- when user keywords are changed, the keywords_data will be updated
- as well as all of its child asset/collections will be updated
- removing assets from a collection will update the asset's collection keywords_data
- deleting a collection will update its asset's collection keywords_data
- changes on keywords_data column will update "searchable_keywords" column
- added "create new collection's" trigger to add collection name into its 'user' keywords_data
- added a new method get_hint_text in simple search wdg
- allows "searchable_keywords" to have a hint_text called "keywords" instead
- if user defined keywords is empty, need to update user keywords_data and searchable keywords too
- changed collection keywords_data, for path to read collection name

### CODE REVIEW

File: src/tactic/command/global_search_trigger.py [hunk 1]
```diff
@@ -104,7 +144,207 @@ def cleanup(my, data):
...
+            user = ""
+            if 'user' in keywords_data:
+                user = keywords_data['user']
+
+            collection = ""
+            if 'collection' in keywords_data:
+                
+                collection_keywords_data = keywords_data['collection']
+                for collection_code in collection_keywords_data.keys():
+                    collection = collection + " " + collection_keywords_data[collection_code]
```
Comment (oboreo) [inline src/tactic/command/global_search_trigger.py]: don't use +  in string concatenation.. keep it as a list and at the next line u can do 
" ".join(collection_list)

File: src/tactic/command/global_search_trigger.py [hunk 2]
```diff
@@ -104,7 +144,207 @@ def cleanup(my, data):
...
+
+            searchable_keywords = path + " " + user + " " + collection
```
Comment (oboreo) [inline src/tactic/command/global_search_trigger.py]: don't use +, use "%s %s %s" %(path, user, " ".join(collection_list))

File: src/tactic/command/global_search_trigger.py [hunk 3]
```diff
@@ -104,7 +144,207 @@ def cleanup(my, data):
...
+        parent_collection_keywords_dict = {}
+
+        # Existing "collection" keywords in child's keywords_data
+        child_keywords_data = child_sobject.get_json_value("keywords_data", {})
+
+        # Existing "collection" keywords in parent's keywords_data
+        parent_keywords_data = parent_sobject.get_json_value("keywords_data", {})
+
+        if 'collection' in child_keywords_data:
+            collection_keywords_dict = child_keywords_data['collection']
```
Comment (oboreo) [inline src/tactic/command/global_search_trigger.py]: use child_keywords_data.get('colleciton')

File: src/tactic/command/global_search_trigger.py [hunk 4]
```diff
@@ -104,7 +144,207 @@ def cleanup(my, data):
...
+
+        if 'collection' in parent_keywords_data:
+            parent_collection_keywords_dict = parent_keywords_data['collection']
```
Comment (oboreo) [inline src/tactic/command/global_search_trigger.py]: same as above

File: src/tactic/command/global_search_trigger.py [hunk 5]
```diff
@@ -104,7 +144,207 @@ def cleanup(my, data):
...
+        results = sql.do_query(statement)
+        for result in results:
+            result = "".join(result)
+            search_codes.append(result)
+
+        return search_codes
+
+    def update_collection_keywords(my, mode, asset_in_asset_sobject):
+        
+        asset_stype = "workflow/asset"
```
Comment (oboreo) [inline src/tactic/command/global_search_trigger.py]: this asset_stype needs to be passed in

File: src/tactic/command/global_search_trigger.py [hunk 6]
```diff
@@ -104,7 +144,207 @@ def cleanup(my, data):
...
+                    child_nest_sobject.set_json_value("keywords_data", child_nest_collection_keywords_data)
+                    child_nest_sobject.commit(triggers=False)
+                    my.set_searchable_keywords(child_nest_sobject)
+
+        sobj.set_json_value("keywords_data", keywords_data)
+        sobj.commit(triggers=False)
+
+    def get_child_codes(my, parent_collection_code):
+
+        from pyasm.biz import Project
```
Comment (oboreo) [inline src/tactic/command/global_search_trigger.py]: The table name asset and asset_in_asset needs to be passed in to this method

File: src/tactic/command/global_search_trigger.py [hunk 7]
```diff
@@ -104,7 +144,207 @@ def cleanup(my, data):
...
+             ng.depth + 1
+            FROM "asset_in_asset" AS r, "asset" AS p1, "asset" AS p2,
+             res AS ng
+            WHERE r."parent_code" = ng."search_code" and depth < 10
+            AND p1."code" = r."parent_code" AND p2."code" = r."search_code"
+            )
+            
+            Select search_code from res;
+            ''' % parent_collection_code
+        else:
```
Comment (oboreo) [inline src/tactic/command/global_search_trigger.py]: these sql statement needs to be defined in database_impl.py and called something else like get_parent_code_cte() and get_child_code_cte()

File: src/tactic/command/global_search_trigger.py [hunk 8]
```diff
@@ -86,6 +93,39 @@ def execute(my):
...
+                keywords_data['user'] = collection_keywords
+                keywords_data['path'] = collection_name
+                sobj.set_json_value("keywords_data", keywords_data)
+                sobj.commit(triggers=False)
+
+            # when user defined keywords column is changed 
+            else:
+                user_keywords = input.get("update_data").get("keywords")
+                if not user_keywords:
+                    user_keywords = ""
```
Comment (oboreo) [inline src/tactic/command/global_search_trigger.py]: so is it going to reset the keywords all the time since the user may not change the keywords every time they edit the sobject

File: src/tactic/command/global_search_trigger.py [hunk 9]
```diff
@@ -32,8 +32,8 @@ def execute(my):
         search_code = input.get('search_code')
 
         sobj_id = input.get('id')
+        sobj = Search.get_by_search_key(search_key)
```
Comment (oboreo) [inline src/tactic/command/global_search_trigger.py]: why is this line moved up here?

Comment (hsuchie4) [inline src/tactic/command/global_search_trigger.py]: resolved

File: src/tactic/command/global_search_trigger.py (hunk 3 above)
Comment (hsuchie4) [inline src/tactic/command/global_search_trigger.py]: fixed

File: src/tactic/command/global_search_trigger.py (hunk 4 above)
Comment (hsuchie4) [inline src/tactic/command/global_search_trigger.py]: fixed

File: src/tactic/command/global_search_trigger.py (hunk 1 above)
Comment (hsuchie4) [inline src/tactic/command/global_search_trigger.py]: fixed

File: src/tactic/command/global_search_trigger.py (hunk 2 above)
Comment (hsuchie4) [inline src/tactic/command/global_search_trigger.py]: fixed

File: src/tactic/command/global_search_trigger.py (hunk 7 above)
Comment (hsuchie4) [inline src/tactic/command/global_search_trigger.py]: fixed

File: src/tactic/command/global_search_trigger.py (hunk 6 above)
Comment (hsuchie4) [inline src/tactic/command/global_search_trigger.py]: fixed

File: src/pyasm/search/database_impl.py [hunk 10]
```diff
@@ -450,7 +450,37 @@ def get_child_cte(my, op_filters):
 
         return stmt
 
- 
+    def get_child_code_cte(my, var_dict):
```
Comment (oboreo) [inline src/pyasm/search/database_impl.py]: it should take in the argument as collection_type, search_type, parent_collection_code.. and you can buidl the var_dict in this method

File: src/pyasm/search/database_impl.py [hunk 11]
```diff
@@ -991,6 +1021,36 @@ def get_child_cte(my, op_filters):
 
         return stmt
 
+    def get_child_code_cte(my, var_dict):
```
Comment (oboreo) [inline src/pyasm/search/database_impl.py]: var_dict should be built in here

File: src/tactic/command/global_search_trigger.py [hunk 12]
```diff
@@ -104,7 +151,170 @@ def cleanup(my, data):
...
+
+        parts = search_type.split("/")
+        collection_type = "%s/%s_in_%s" % (parts[0], parts[1], parts[1])
+
+        # Check if connection between asset and asset_in_asset is in place
+        if collection_type not in SearchType.get_related_types(search_type):
+            return search_codes
+
+        var_dict = {
+            'parent_collection_code': parent_collection_code,
```
Comment (oboreo) [inline src/tactic/command/global_search_trigger.py]: pass in these 3 arguments to the method get_child_code_cte

File: src/tactic/command/global_search_trigger.py [hunk 13]
```diff
@@ -45,7 +46,15 @@ def execute(my):
             search_code = sobj_id
 
         search_type = SearchKey.extract_search_type(search_key)
-        
+
+        input_search_type = input.get("search_type")
+        base_search_type = input_search_type.split("?")[0]
+
```
Comment (oboreo) [inline src/tactic/command/global_search_trigger.py]: initialize with 
has_keywords_data = False

File: src/tactic/command/global_search_trigger.py [hunk 14]
```diff
@@ -86,6 +104,77 @@ def execute(my):
         data.update( my.cleanup(caller.get_value("description", no_exception=True) ))
         data.update( my.cleanup(caller.get_value("keywords", no_exception=True) ))
 
+
+        # If keywords_data column exists and collection is being changed
+        if has_keywords_data:
+
```
Comment (oboreo) [inline src/tactic/command/global_search_trigger.py]: intiialize 
rename_collection = False

File: src/tactic/command/global_search_trigger.py [hunk 15]
```diff
@@ -86,6 +104,77 @@ def execute(my):
...
+                        user = user.replace(old_collection_name, collection_name)
+                        keywords_data['user'] = user
+                        
+                        sobj.set_json_value("keywords_data", keywords_data)
+                        sobj.commit(triggers=False)
+
+                        my.update_user_keywords(sobj, user, base_search_type)
+                        my.set_searchable_keywords(sobj)
+
+                # If user defined keywords column is changed
```
Comment (oboreo) [inline src/tactic/command/global_search_trigger.py]: change the comment to user_keywords

File: src/tactic/command/global_search_trigger.py [hunk 16]
```diff
@@ -86,6 +104,77 @@ def execute(my):
...
+
+                        my.update_user_keywords(sobj, user, base_search_type)
+                        my.set_searchable_keywords(sobj)
+
+                # If user defined keywords column is changed 
+                elif has_user_keywords:
+                    user_keywords = input.get("update_data").get("user_keywords")
+
+                    if not user_keywords:
+                        user_keywords = ""
```
Comment (oboreo) [inline src/tactic/command/global_search_trigger.py]: why is these 2 lines if not user_keywords, user_keywords = '' unique to collection?

File: src/tactic/command/global_search_trigger.py [hunk 17]
```diff
@@ -86,6 +104,77 @@ def execute(my):
...
+            # If regular asset keywords being changed
+            else:
+                if has_user_keywords:
+                    user_keywords = input.get("update_data").get("user_keywords")
+
+                    my.update_user_keywords(sobj, user_keywords, base_search_type)
+                    my.set_searchable_keywords(sobj)
+                
+        # Collection relationships being created or added
+        stype_obj = SearchType.get(base_search_type)
```
Comment (oboreo) [inline src/tactic/command/global_search_trigger.py]: this section looks very much like the delete mode logic.. pls put it in a method and call it here and the delete section so the code is not duplicated twice.
